flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java

```diff
@@ -218,26 +221,57 @@ public void run() {
 				}
 			}
 
-			// seek the consumer to the initial offsets
+			List<KafkaTopicPartition> partitionsWithNoOffset = new ArrayList<>();
 			for (KafkaTopicPartitionState<TopicPartition> partition : subscribedPartitions()) {
 				if (partition.isOffsetDefined()) {
 					LOG.info("Partition {} has restored initial offsets {} from checkpoint / savepoint; seeking the consumer " +
 						"to position {}", partition.getKafkaPartitionHandle(), partition.getOffset(), partition.getOffset() + 1);
 
 					consumer.seek(partition.getKafkaPartitionHandle(), partition.getOffset() + 1);
 				} else {
-					// for partitions that do not have offsets restored from a checkpoint/savepoint,
-					// we need to define our internal offset state for them using the initial offsets retrieved from Kafka
-					// by the KafkaConsumer, so that they are correctly checkpointed and committed on the next checkpoint
+					partitionsWithNoOffset.add(partition.getKafkaTopicPartition());
+				}
+			}
 
-					long fetchedOffset = consumer.position(partition.getKafkaPartitionHandle());
+			if (partitionsWithNoOffset.size() == subscribedPartitions().length) {
+				// if all partitions have no initial offsets, that means we're starting fresh
+				switch (startupMode) {
+					case EARLIEST:
+						LOG.info("Setting starting point as earliest offset for partitions {}", partitionsWithNoOffset);
+
+						seekPartitionsToBeginning(consumer, convertKafkaPartitions(subscribedPartitions()));
```

Comment:
Its a bit inefficient to convert the `subscribedPartitions()` to an ArrayList, and then in `seekPartitionsToBeginning` the List is converted back into an array. I think we can save the `ArrayList` step and create an array immediately.